core/network/src/protocol.rs

```diff
@@ -676,21 +694,48 @@ impl<B: BlockT, S: NetworkSpecialization<B>, H: ExHashT> Protocol<B, S, H> {
 
 			let cache_limit = NonZeroUsize::new(1_000_000).expect("1_000_000 > 0; qed");
 
-			let peer = Peer {
-				info: PeerInfo {
-					protocol_version: status.version,
-					roles: status.roles,
-					best_hash: status.best_hash,
-					best_number: status.best_number
+			let info = match self.handshaking_peers.remove(&who) {
+				Some(hanshaking) => {
+					let peer_id = Some(hanshaking.peer_id);
+					let peer_info = PeerInfo {
+						peer_id: peer_id.clone(),
+						protocol_version: status.version,
+						roles: status.roles,
+						best_hash: status.best_hash,
+						best_number: status.best_number
+					};
+					self.connected_peers
+						.write()
+						.insert(who, ConnectedPeer { peer_id, peer_info: peer_info.clone() });
+					peer_info
+				},
+				None => {
```

Comment:
This statement should be unreachable so I think we should just log the error and return.